Better `pkgs/by-name` errors, minor fixes and improvements

infinisil opened the pull request:
## Description of changes

This is a bit of a random assortment of `pkgs/by-name` improvements:
- Better errors in many cases. In particular errors when a package in `pkgs/by-name` is overridden with a manual definition in `all-packages.nix`, but it's not of the expected form.
- A more spaced-out error formatting for some errors (could do all in the future)
- A minor false negative fix, for when a package in `pkgs/by-name` is overridden with a `callPackage` whose first argument is not a path. This also makes the code easier to understand.
- A low-hanging performance improvement by running the check on two cores, which should shave around 25 seconds from CI.
- Indications about whether the PR would break master, fixes master, or doesn't affect it when merged. This should make sure that we don't run into cases like https://github.com/NixOS/nixpkgs/issues/289649 again. This essentially re-implements https://github.com/NixOS/nixpkgs/pull/257735, but now in Rust

I fully admit that the code isn't the best, but I've been getting side-tracked trying to improve it for like 2 days :upside_down_face: 

This relates to https://github.com/NixOS/nixpkgs-check-by-name/issues/6, more is needed to address that, probably just more docs, but that can be done separately

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

## Things done

- [x] Updated/added tests

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
tests.nixpkgs-check-by-name: Improve inherit detection

Detect inherit's better, such that a future commit can use this
information in an error message

infinisil committed:
tests.nixpkgs-check-by-name: Fix allowing non-path overrides

An edge case was allowed when it shouldn't be: A package defined in
`pkgs/by-name` could be overridden in `all-packages.nix` if it was of
the form `callPackage (<expr>) { <non-empty> }`.

This is not right, it's required that the first argument be the path
matching the package to be overridden.

infinisil committed:
tests.nixpkgs-check-by-name: Better error for non-syntactic callPackage

infinisil committed:
tests.nixpkgs-check-by-name: Improve error for wrong package file

infinisil committed:
tests.nixpkgs-check-by-name: Improve non-syntactic callPackage error more

infinisil committed:
tests.nixpkgs-check-by-name: Improve more errors

infinisil committed:
tests.nixpkgs-check-by-name: Improve errors relating to the base branch

infinisil committed:
tests.nixpkgs-check-by-name: Evaluate base and main branch in parallel

infinisil committed:
tests.nixpkgs-check-by-name: More consistent errors

infinisil committed:
tests.nixpkgs-check-by-name: Better error for empty second arg

infinisil committed:
tests.nixpkgs-check-by-name: More spaces in errors

infinisil committed:
tests.nixpkgs-check-by-name: Improve errors for new packages

Or rather, make it more consistent

infinisil committed:
tests.nixpkgs-check-by-name: Fix internal docs

philiptaron reviewed (changes requested):
The tests not being reproducible outside your enlistment (`/home/tweagysil/src`) looks like a blocker. There's also just a lot of English verbiage suggestions that in total add up to "request changes." I'm not tied to any particular suggestion, so if you see a better way of saying it, please do so immediately.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -55,7 +90,7 @@ pub enum NixpkgsProblem {
         relative_package_dir: PathBuf,
         subpath: PathBuf,
-        io_error: io::Error,
+        io_error: String,
```
I changed this because I had to make `NixpkgsProblem` a `Clone`, because of this new use:

https://github.com/NixOS/nixpkgs/blob/64da6178bf10b92f57352d7d0cfca7eebbd527df/pkgs/test/nixpkgs-check-by-name/src/ratchet.rs#L129-L139

And I can't find a good way to avoid that just yet

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/references.rs:
```diff
@@ -66,7 +66,7 @@ fn check_path(
                 relative_package_dir: relative_package_dir.to_path_buf(),
                 subpath: subpath.to_path_buf(),
-                io_error,
+                io_error: io_error.to_string(),
```
(see https://github.com/NixOS/nixpkgs/pull/290743#discussion_r1503258930)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -276,51 +292,86 @@ fn by_name(
+                        let relative_file =
+                            location.relative_file(nixpkgs_path).with_context(|| {
+                                format!(
+                                    "Failed to resolve location file of attribute {attribute_name}"
```
I like "defined", will change :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -77,34 +73,67 @@ fn main() -> ExitCode {
+            for error in errors {
+                writeln!(error_writer, "{}", error.to_string().red())?
+            }
+            writeln!(error_writer, "{}", "The base branch is broken and still has above problems with this PR, these need to be fixed first.\nConsider reverting the PR that introduced these problems in order to prevent more failures of unrelated PRs.".yellow())?;
```
I think it works like this:
```suggestion
            writeln!(error_writer, "{}", "The base branch is broken and still has above problems with this PR, which need to be fixed first.\nConsider reverting the PR that introduced these problems in order to prevent more failures of unrelated PRs.".yellow())?;
```

philiptaron replied:
Yes, that works.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -77,34 +73,67 @@ fn main() -> ExitCode {
+    match (base_result, main_result) {
+        (Failure(_), Failure(errors)) => {
+            // Base branch fails and the PR doesn't fix it and may also introduce additional problems
+            for error in errors {
+                writeln!(error_writer, "{}", error.to_string().red())?
+            }
```
In the future I'd like to move towards a tree-based structure for representing checking context and their results. Currently that already somewhat happens for the successful case with the ratchets, but there's no reason this couldn't also be done for the failing case in a more generic way. And once that exists, it's going to be easy to determine new errors.

So I'd like to keep it as is for now :D

philiptaron replied:
Sure, sounds good.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -77,34 +73,67 @@ fn main() -> ExitCode {
+            writeln!(
+                error_writer,
+                "{}",
+                "The base branch was broken, but this PR fixes it, nice job!".green()
```
I'd like to avoid `;` because I don't think it's ever needed :P

I think this is okay:
```suggestion
                "The base branch is broken, but this PR fixes it. Nice job!".green()
```

philiptaron replied:
The Ernest Hemingway option is to always use periods. 

Semicolons are the cheater's way out of a comma splice; since they act as a both a period and a comma, one can replace the splicing comma with a semicolon with no other edits, and the "flow" intention of using the comma is preserved.

But all sorts of edits are possible, and this is a fine one. 🍶

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -77,34 +73,67 @@ fn main() -> ExitCode {
+            writeln!(
+                error_writer,
+                "{}",
+                "This PR introduces the above problems, merging would break the base branch"
```
I'll go with
```suggestion
                "This PR introduces the problems listed above. Please fix them before merging, otherwise the base branch would break."

```

I think it's important to mention that the base branch would break to really highlight that merging is not okay.

philiptaron replied:
Sounds good.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -77,34 +73,67 @@ fn main() -> ExitCode {
+                    for error in errors {
+                        writeln!(error_writer, "{}", error.to_string().red())?
+                    }
+                    writeln!(error_writer, "{}", "This PR introduces the above problems compared to the base branch, merging is discouraged, but would not break the base branch".yellow())?;
```
Hmm I don't think that's entirely right. The failure is specifically in the PR itself, the transition from the base branch to the PRs branch. The branches by themselves don't contain such errors.

Instead for now I'll go with
```suggestion
                    writeln!(error_writer, "{}", "This PR introduces additional instances of discouraged patterns as listed above. Merging is discouraged but would not break the base branch.".yellow())?;
```

philiptaron replied:
That's substantially clearer than the original. 👍🏻

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -87,8 +86,9 @@ pub struct CallPackageArgumentInfo {
-    /// If the location is not of the form `<attr> = callPackage <arg1> <arg2>;`, `None` is
-    /// returned.
+    /// If the location is not of the form `<attr> = callPackage <arg1> <arg2>;`, `Ok((None, String))` is
+    /// returned, with String being how the definition looks like.
```
Changing it to

> If the definition at the given location is not of the form `<attr> = callPackage <arg1> <arg2>;`, `Ok((None, String))` is returned, with `String` being the definition itself.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -119,19 +122,24 @@ impl NixFile {
         line: usize,
         column: usize,
-    ) -> anyhow::Result<Option<ast::AttrpathValue>> {
+    ) -> anyhow::Result<Result<ast::AttrpathValue, String>> {
```
`Either` sounds fine to me for now, especially because it's internal to the module and not `pub`

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -40,9 +43,41 @@ pub enum NixpkgsProblem {
-    WrongCallPackage {
-        relative_package_file: PathBuf,
+    EmptyArgument {
+        package_name: String,
+        file: PathBuf,
+        line: usize,
+        column: usize,
+        definition: String,
```
That's one of the main reasons I want to go for some generic tree-like context representation as mentioned in https://github.com/NixOS/nixpkgs/pull/290743#discussion_r1508222275. Am eyeing this for the future :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -158,12 +195,101 @@ impl fmt::Display for NixpkgsProblem {
+
+                        {package_name} = callPackage ./{} {{ /* ... */ }};
+
+                      Notably the second argument must not be empty, which is not the case.
+                      It is defined in {}:{} as
```
I'll apply the rewording.

Regarding what to do next, I'd like to push that to another PR. I think it would be best to have documentation for resolving each of these errors and then linking to it from the errors.

infinisil replied:
Though for this specific error it's a fairly straightforward improvement, which I now implemented with https://github.com/NixOS/nixpkgs/pull/290743/commits/7858f062884e2318dda212ab7cc1ceda871195b2 :)

philiptaron replied:
Nice! I like it a lot.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -158,12 +195,101 @@ impl fmt::Display for NixpkgsProblem {
+
+                        {package_name} = callPackage {} {{ /* ... */ }};
+
+                      This is however not the case: The first `callPackage` argument is the wrong path.
+                      It is defined in {}:{}:{} as
```
I'll just use

```suggestion
                      However, in this PR, the first `callPackage` argument is the wrong path. See the definition in {}:{}:
```

Since the wrong path is more easily visible from the definition in the next line

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -248,13 +382,18 @@ impl fmt::Display for NixpkgsProblem {
                 let extra =
                     if *empty_arg {
-                        "Since the second `callPackage` argument is `{ }`, no manual `callPackage` (e.g. in `pkgs/top-level/all-packages.nix`) is needed anymore."
+                        format!("Since the second `callPackage` argument is `{{ }}`, no manual `callPackage` in {} is needed anymore.", file.display())
```
Not entirely sure how to go about that. Maybe better as a separate PR in the future :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -271,3 +410,30 @@ impl fmt::Display for NixpkgsProblem {
+
+/// Creates a Nix path expression that when put into Nix file `from_file`, would point to the `to_file`.
+fn create_path_expr(from_file: impl AsRef<Path>, to_file: impl AsRef<Path>) -> String {
+    // FIXME: Clean these unwrap's up
+    // But these should never trigger because we only call this function with relative paths, and only
+    // with `from_file` being an actual file (so there's always a parent)
+    let from = RelativePath::from_path(&from_file)
+        .unwrap()
+        .parent()
+        .unwrap();
+    let to = RelativePath::from_path(&to_file).unwrap();
```
Sounds good to me!

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:
```diff
@@ -248,13 +382,18 @@ impl fmt::Display for NixpkgsProblem {
-                    "pkgs.{package_name}: This is a new top-level package of the form `callPackage {call_package_arg} {{ }}`. Please define it in {} instead. See `pkgs/by-name/README.md` for more details. {extra}",
+                    "
+                    - Attribute `pkgs.{package_name}` is a new top-level package using `pkgs.callPackage {call_package_arg} {{ /* ... */ }}`.
+                      Please define it in {} instead.
+                      See `pkgs/by-name/README.md` for more details.
+                      {extra}
```
Agreed, doing that!

infinisil committed:
tests.nixpkgs-check-by-name: Various minor improvements

Co-Authored-By: Philip Taron <philip.taron@gmail.com>

infinisil committed:
tests.nixpkgs-check-by-name: Fix absolute path in errors

infinisil committed:
tests.nixpkgs-check-by-name: Better empty argument error

infinisil committed:
tests.nixpkgs-check-by-name: Use RelativePath for relative paths

Makes the code easier to understand and less error-prone

infinisil committed:
tests.nixpkgs-check-by-name: More inline format! arguments

Now that the previous commit removed all the .display()'s that were
previously necessary for PathBuf's, but now aren't for RelativePathBuf,
we can also inline the format! arguments

infinisil commented on pkgs/test/nixpkgs-check-by-name/tests/sorted-order/expected:
```diff
@@ -1,4 +1,29 @@
+- Attribute `pkgs.d` is a new top-level package using `pkgs.callPackage ... { /* ... */ }`.
+  Please define it in pkgs/by-name/d/d/package.nix instead.
+  See `pkgs/by-name/README.md` for more details.
+  Since the second `callPackage` argument is `{ }`, no manual `callPackage` in /home/tweagysil/src/nixpkgs/by-name-improv/pkgs/test/nixpkgs-check-by-name/tests/sorted-order/all-packages.nix is needed anymore.
```
See https://github.com/NixOS/nixpkgs/pull/290743#discussion_r1508242997

infinisil commented on pkgs/test/nixpkgs-check-by-name/tests/move-to-non-by-name/expected:
```diff
@@ -1,4 +1,13 @@
-pkgs.foo2: This top-level package was previously defined in pkgs/by-name/fo/foo2/package.nix, but is now manually defined as `callPackage ./without-config.nix { }` (e.g. in `pkgs/top-level/all-packages.nix`). Please move the package back and remove the manual `callPackage`.
-pkgs.foo3: This top-level package was previously defined in pkgs/by-name/fo/foo3/package.nix, but is now manually defined as `callPackage ... { ... }` (e.g. in `pkgs/top-level/all-packages.nix`). While the manual `callPackage` is still needed, it's not necessary to move the package files.
-pkgs.foo4: This top-level package was previously defined in pkgs/by-name/fo/foo4/package.nix, but is now manually defined as `callPackage ./with-config.nix { ... }` (e.g. in `pkgs/top-level/all-packages.nix`). While the manual `callPackage` is still needed, it's not necessary to move the package files.
+- Attribute `pkgs.foo1 was previously defined in pkgs/by-name/fo/foo1/package.nix, but is now manually defined as `callPackage ... { /* ... */ }` in /home/tweagysil/src/nixpkgs/by-name-improv/pkgs/test/nixpkgs-check-by-name/tests/move-to-non-by-name/all-packages.nix.
```
Now fixed with https://github.com/NixOS/nixpkgs/pull/290743/commits/f056449c462ff6c5f933f3c297d605ff02147a99

infinisil commented:
@philiptaron Thanks for the extensive review! I either addressed or implemented all of them now. I was also able to do a nice refactoring using `RelativePath[Buf]`

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -339,7 +371,7 @@ pub enum ResolvedPath {
     Outside,
     /// The path is within the given absolute path.
     /// The `PathBuf` is the relative path under the given absolute path.
```
comment now out of date

philiptaron reviewed (approved):
Tests and messages look good.
